File: Golang/Data Structures/Linked List/singly-linked-list.go
type Node struct {
	Value int
	Next  *Node
}

type List struct {
	Head *Node
	Tail *Node
	Size int
}

func (list *List) Init() *List {

	ll := &List{nil, nil, 0}

	return ll
}

func (list *List) size() int {
	return list.Size
}

func (list *List) print() {
	current := list.Head
	if list.Size == 0 {
		return
	}
	for current != nil {
		fmt.Printf("%v ", current.Value)
		current = current.Next
	}
}

func (list *List) prepend(val int) {
	node := &Node{val, nil}

	if list.Head == nil {
		list.Head = node
		list.Tail = node
	} else {
		node.Next = list.Head
		list.Head = node
	}

	list.Size++
}

func (list *List) append(val int) {
	node := &Node{val, nil}

	if list.Size == 0 {
		list.Head = node
		list.Tail = node
	} else {
		list.Tail.Next = node
		list.Tail = list.Tail.Next
	}

	list.Size++
}

func (list *List) popLeft() {

	if list.Size > 0 {
		list.Head = list.Head.Next
		list.Size--
	}
}

func (list *List) popRight() {

	if list.Size > 0 {
		current := list.Head

		for current.Next != list.Tail {
			current = current.Next
		}
		list.Tail = current
		current.Next = nil
		list.Size--
	}
}

func (list *List) insert(value, index int) {

	if index > list.Size {
		fmt.Println("ERROR: Index of insert poisiton must be less than the size of the Linked List")
	}

	if index < 0 {
		fmt.Println("Error: Index of insert position must be greater than 0")
	}

	if index == 0 {
		list.prepend(value)
	} else if index == list.Size {
		list.append(value)
	} else {
		current := list.Head
		i := 0
		for i < index-1 {
			i++
			current = current.Next
		}
		node := &Node{value, nil}
		nextNode := current.Next
		node.Next = nextNode
		current.Next = node
		list.Size++
	}
}

func (list *List) remove(index int) {

	if index >= list.Size {
		fmt.Println("ERROR: Index must be less than the size of Linked List")
	}

	if index < 0 {
		fmt.Println("ERROR: Index must be greater than 0")
	}

	if index == 0 {
		list.popLeft()
	} else if index == list.Size-1 {
		list.popRight()
	} else {

		current := list.Head
		i := 0
		for i < index-1 {
			current = current.Next
			i++
		}

		current.Next = current.Next.Next
		list.Size--
	}

}

func (list *List) clear() {
	list.Head = nil
	list.Tail = nil
	list.Size = 0
}

File: Java/Data Structures/Linked Lists/SinglyLinkedList.java
package com.believer.LinkedList;

class Node <T> {
    protected T value;
    protected Node<T> next;

    public Node(T value) {
        this.value = value;
        this.next = null;
    }
}

public class SinglyLinkedList <T> {

    private Node<T> head;
    private Node<T> tail;
    private int size;

    public SinglyLinkedList() {
        this.head = null;
        this.tail = null;
        this.size = 0;
    }

    public void print() {

        Node<T> current = this.head;

        while (current != null) {
            if (current.next == null) {
                System.out.print(current.value);
            } else {
                System.out.print(current.value + " -> ");
            }
            current = current.next;
        }
    }

    public void prepend(T value) {

        Node<T> node = new Node<>(value);

        if (this.head == null) {
            this.head = node;
            this.tail = node;
        } else {
            node.next = this.head;
            this.head = node;
        }

        this.size++;
    }

    public void append(T value) {

        Node<T> node = new Node<>(value);

        if (this.head == null) {
            this.head= node;
            this.tail = node;
        } else {
            this.tail.next = node;
            this.tail = this.tail.next;
        }

        this.size++;
    }

    public T popLeft() throws Exception {

        T popValue = null;

        if (this.head == null) {
            throw new Exception("Cannot pop from empty list");
        } else if (this.head.next == null) {
            popValue = this.head.value;
            this.head = null;
            this.tail = null;
        } else {
            popValue = this.head.value;
            this.head = this.head.next;
        }

        this.size--;

        return popValue;
    }

    public T popRight() throws Exception {


        if (this.head == null) {
            throw new Exception("Cannot pop from empty list");
        }

        T popValue = this.tail.value;

        if (this.head.next == null) {
            this.head = null;
            this.tail = null;
        } else {

            Node<T> current = this.head;

            while (current.next != this.tail) {
                current = current.next;
            }

            this.tail = current;
            this.tail.next = null;
        }

        this.size--;

        return popValue;
    }

    public void remove(int index) throws Exception {

        if (this.head == null) {
            throw new Exception("Cannot remove from empty list");
        }

        if (index < 0 || index > this.size-1) {
            throw new Exception("Invalid index");
        }

        if (index == 0) {
            this.popLeft();
        } else if (index == this.size-1) {
            this.popRight();
        } else {

            Node<T> current = this.head;
            int tempIndex = 0;

            while (tempIndex < index-1) {
                current = current.next;
                tempIndex++;
            }

            current.next = current.next.next;

            this.size--;
        }
    }

    public void remove(T value) throws Exception {

        if (this.head == null) {
            throw new Exception("Cannot remove from empty list");
        }

        if (this.head.value == value) {
            this.popLeft();
        } else if (this.tail.value == value) {
            this.popRight();
        } else {

            Node<T> current = this.head;

            while (current.next.value != value) {
                current = current.next;
            }

            current.next = current.next.next;

            this.size--;
        }
    }

    public void clear() {
        this.head = null;
        this.tail = null;
        this.size = 0;
    }

}
